### PR TITLE
[CBR-96] benchmarks

### PR DESCRIPTION
**Pull request by coot**
## Description

This PR includes:
* benchamarks for
  - `verifyAndApplyBlocks`
  - `verifyBlocksPrefix` (without application)
  - `verifyHeaders`
* an executable that I used to profile the verification and application of blocks.

I have more that comes on top of this PR, let's first try to merge this.

This PR required some refactoring of the `verifyBlocksPrefix` function so that we can verify blocks without applying them.  In addition I pulled some code from various parts of verification into `VerifyBlocksContext`: current slot, `BlockVersion` and `BlockVersionData`.

I also included here a commit which added `--nix` argument to `scripts/launchdemo.sh` which is passed to `stack`.

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CBR-96

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: generator/src/Test/Pos/Block/Logic/Mode.hs [hunk 1]
```diff
@@ -505,9 +505,9 @@ instance MonadBListener BlockTestMode where
 
 type instance MempoolExt BlockTestMode = EmptyMempoolExt
 
-instance HasConfigurations => MonadTxpLocal (BlockGenMode EmptyMempoolExt BlockTestMode) where
-    txpNormalize = withCompileInfo def $ txNormalize
-    txpProcessTx = withCompileInfo def $ txProcessTransactionNoLock
+instance MonadTxpLocal (BlockGenMode EmptyMempoolExt BlockTestMode) where
+    txpNormalize _ = return () -- withCompileInfo def $ txNormalize
+    txpProcessTx _ _ = return $ Right () -- withCompileInfo def $ txProcessTransactionNoLock
```
Comment (parsonsmatt) [inline generator/src/Test/Pos/Block/Logic/Mode.hs]: This change seems odd to me

File: lib/bench/Bench/Pos/Diffusion/BlockDownload.hs [hunk 2]
```diff
@@ -287,14 +287,6 @@ runBlockDownloadBenchmark :: Criterion.Mode -> NodeId -> (Int -> IO ()) -> Diffu
 runBlockDownloadBenchmark mode serverAddress setStreamIORef client =
     Criterion.runMode mode $ blockDownloadBenchmarks serverAddress setStreamIORef client
 
--- It's surprisingly cumbersome to give a non-orphan 'NFData' instance on
--- 'BlockHeader', since we have that 'Blockchain' typeclass with a bunch of
--- data families. 'BHeaderHash GenesisBlockchain', for instance, must have
--- an 'NFData' instance, but in that module we don't yet know that this is
--- in fact 'HeaderHash' ~ 'Crypto.Digest Blake2b_256'.
--- Anyway, there's a whole saga of pain caused by that silly abstraction.
-instance NFData BlockHeader
```
Comment (parsonsmatt) [inline lib/bench/Bench/Pos/Diffusion/BlockDownload.hs]: :joy:

File: generator/src/Test/Pos/Block/Logic/Mode.hs (hunk 1 above)
Comment (coot) [reply generator/src/Test/Pos/Block/Logic/Mode.hs]: Good catch, I'll revert this.

Comment (coot) [reply generator/src/Test/Pos/Block/Logic/Mode.hs]: Done

**Comment by coot** [comment]
@parsonsmatt I addressed your review and I restored the instance.

**Comment by coot** [comment]
After rebasing I started to get:
```
cardano-sl-verification-bench: genTransaction@txProcessTransaction: got left: input is not a known unspent input: TxInUtxo 053377ea #0
CallStack (from HasCallStack):
  error, called at src/Universum/Debug.hs:60:11 in universum-1.1.1-Auxpv5XhG5VAhqzacQVnT8:Universum.Debug
  error, called at src/Pos/Generator/Block/Payload.hs:227:24 in cardano-sl-generator-1.3.0-GhPxCjUaOkbHmzfWnRTquK:Pos.Generator.Block.Payload
```
Let me fix this before merging.

**Comment by coot** [comment]
When I run a profiling executable which is part of this PR, it gives a slightly better info in what's going on:
```
stack --nix exec cardano-sl-verification-bench-exe -- --config lib/configuration.yaml --block-cache /tmp/block.cache
[verification-bench:INFO] using configurations: ConfigurationOptions {cfoFilePath = "lib/configuration.yaml", cfoKey = "bench-validation", cfoSystemStart = Just 1530721078837497, cfoSeed = Nothing}
[*production*:DEBUG] Loading SSC global state
[*production*:INFO] Loaded SSC state:   certificates from: [df4901e8f6abd96a, f83e9e0270abcc74, c96ac1e118c0d384, 81713ef44a946310]

[testing:INFO] Generating blocks
[testing:DEBUG] Loading SSC global state
[testing:INFO] Loaded SSC state:   certificates from: [df4901e8f6abd96a, f83e9e0270abcc74, c96ac1e118c0d384, 81713ef44a946310]

[testing:DEBUG] Transaction is processed successfully: f962ca8382c86281
[testing:INFO] We are trying to create main block, our tip header is
GenesisBlockHeader:
    hash: 7ae1defcca0971a422430b34f9b93ba8fba818f8e504de818087aaa0add429a8
    previous block: 791f4256e14c67b9035c3b80a0826adf719d3636c18eef16c98b84b833723d51
    epoch: #0
    difficulty: 0

[testing:INFO] Created main block of size: 960 B
[testing:INFO] We are trying to create main block, our tip header is
MainBlockHeader:
    hash: fbae2a2e7a6fd5a1612bb8bc056989dd344bad289235231b4af4a08431f02538
    previous block: 7ae1defcca0971a422430b34f9b93ba8fba818f8e504de818087aaa0add429a8
    slot: 0th slot of 0th epoch
    difficulty: 1
    leader: pub:440b6491
    signature: BlockPSignatureHeavy: Proxy signature { psk = ProxySk { w = #0, iPk = pub:440b6491, dPk = pub:ba5b2a37 } }
    block: v0.0.0
    software: cardano-sl:0

[testing:WARNING] txGetPayload: tip mismatch (in DB: )fbae2a2e7a6fd5a1, (in mempool: 7ae1defcca0971a4)
[testing:WARNING] US payload can't be created due to tip mismatch (our payload is for 7ae1defc, but we want to create payload based on tip fbae2a2e)
[testing:INFO] Created main block of size: 665 B
[testing:DEBUG] Transaction is processed successfully: c50c1a798e5ed99c
[testing:DEBUG] Transaction is processed successfully: 78ca18fce0376e70
[testing:INFO] We are trying to create main block, our tip header is
MainBlockHeader:
    hash: d8f2f6581dfd42c8ef89448f29ca52e7c22ca791b1c5e6288e2521b9dd2d9e6a
    previous block: fbae2a2e7a6fd5a1612bb8bc056989dd344bad289235231b4af4a08431f02538
    slot: 1st slot of 0th epoch
    difficulty: 2
    leader: pub:a3219c1a
    signature: BlockPSignatureHeavy: Proxy signature { psk = ProxySk { w = #0, iPk = pub:a3219c1a, dPk = pub:65cad307 } }
    block: v0.0.0
    software: cardano-sl:0

[testing:WARNING] txGetPayload: tip mismatch (in DB: )d8f2f6581dfd42c8, (in mempool: 7ae1defcca0971a4)
[testing:WARNING] US payload can't be created due to tip mismatch (our payload is for 7ae1defc, but we want to create payload based on tip d8f2f658)
[testing:INFO] Created main block of size: 665 B
[testing:DEBUG] Transaction processing failed: 0864e06088991c10
cardano-sl-verification-bench-exe: genTransaction@txProcessTransaction: got left: input is not a known unspent input: TxInUtxo 517f70a4 #0
CallStack (from HasCallStack):
  error, called at src/Universum/Debug.hs:60:11 in universum-1.1.1-Auxpv5XhG5VAhqzacQVnT8:Universum.Debug
  error, called at src/Pos/Generator/Block/Payload.hs:227:24 in cardano-sl-generator-1.3.0-3uBXyAHQvHhCpoiyaZsC5Q:Pos.Generator.Block.Payload
```

**Comment by coot** [comment]
@erikd I will close this for now and push smaller PRs which are pre requirements for this one, we'll get back to this when we're done with those.

**Pull request closed without merging**